src/rgw/driver/d4n/rgw_sal_d4n.cc

```diff
@@ -352,60 +404,501 @@ int D4NFilterObject::D4NFilterReadOp::prepare(optional_yield y, const DoutPrefix
+	  } else {
+	    ldpp_dout(dpp, 20) << "D4N Filter: Block directory set operation succeeded." << dendl;
+	  }
+	}
+      }
+    }
+  }
+
+  if (ret < 0) 
+    ldpp_dout(dpp, 20) << "D4N Filter: Cache iterate operation failed." << dendl;
+
+  return next->iterate(dpp, ofs, end, cb, y); */
+}
+
+int D4NFilterObject::D4NFilterReadOp::D4NFilterGetCB::flush_last_part()
+{
+  last_part = true;
+  return handle_data(bl_rem, 0, bl_rem.length());
+}
+
+int D4NFilterObject::D4NFilterReadOp::D4NFilterGetCB::handle_data(bufferlist& bl, off_t bl_ofs, off_t bl_len)
+{
+  auto rgw_get_obj_max_req_size = g_conf()->rgw_get_obj_max_req_size;
+
+  if (!last_part && bl.length() <= rgw_get_obj_max_req_size) {
+    auto r = client_cb->handle_data(bl, bl_ofs, bl_len);
+
+    if (r < 0) {
+      return r;
+    }
+  }
+
+  //Accumulating data from backend store into rgw_get_obj_max_req_size sized chunks and then writing to cache
+  if (write_to_cache) {
+    const std::lock_guard l(d4n_get_data_lock);
+    rgw::d4n::CacheBlock block;
+    rgw::d4n::BlockDirectory* blockDir = source->driver->get_block_dir();
+    block.hostsList.push_back(blockDir->get_addr().host + ":" + std::to_string(blockDir->get_addr().port));
+    block.cacheObj.bucketName = source->get_bucket()->get_name();
+    block.cacheObj.objName = source->get_key().get_oid();
+
+    if (bl.length() > 0 && last_part) { // if bl = bl_rem has data and this is the last part, write it to cache
+      std::string oid = this->oid + "_" + std::to_string(ofs) + "_" + std::to_string(bl_len);
+      block.size = bl.length();
+      block.blockId = ofs;
+      uint64_t freeSpace = filter->get_cache_driver()->get_free_space(dpp);
+      while(freeSpace < block.size) {
+        freeSpace += filter->get_policy_driver()->cachePolicy->eviction(dpp, filter->get_cache_driver());
+      }
+      if (filter->get_cache_driver()->put_async(dpp, oid, bl, bl.length(), source->get_attrs()) == 0) {
+        filter->get_policy_driver()->cachePolicy->update(dpp, oid, ofs, bl.length(), filter->get_cache_driver());
+        /* Store block in directory */
+        if (!blockDir->exist_key(oid)) {
+          int ret = blockDir->set_value(&block);
+          if (ret < 0) {
+            ldpp_dout(dpp, 0) << "D4N Filter: Block directory set operation failed." << dendl;
+            return ret;
+          } else {
+            ldpp_dout(dpp, 20) << "D4N Filter: Block directory set operation succeeded." << dendl;
+          }
+        }
+      }
+    } else if (bl.length() == rgw_get_obj_max_req_size && bl_rem.length() == 0) { // if bl is the same size as rgw_get_obj_max_req_size, write it to cache
+      std::string oid = this->oid + "_" + std::to_string(ofs) + "_" + std::to_string(bl_len);
+      ofs += bl_len;
+      block.blockId = ofs;
+      block.size = bl.length();
+      uint64_t freeSpace = filter->get_cache_driver()->get_free_space(dpp);
+      while(freeSpace < block.size) {
```

Comment:
Are we only checking for the block size and not adding the attribute size here because the attributes are being stored for the entire object rather than just the block?

Reply:
Correct. But are you concerned about the `localWeight` attribute that gets added as part of the `update` call?

Reply:
But we had decided to move this check to `eviction` call correct?

Reply:
The available cache space is calculated in the `eviction` call and yes, the `localWeight` attribute will also take up a slight amount of space that must be considered during eviction. This should account for both the attribute name and value size, correct? With the attribute name including the `user.rgw.` prefix?

Reply:
Update: Due to a recent discussion on this subject, we will keep the block size as the size to free up in the `eviction` call.